woocommerce/woocommerce
# Run performance metrics on pull request

### Changes proposed in this Pull Request:

This PR introduces delta calculations for all the performance metrics (frontend, editor, and product editor) so that we can see changes in percentage.

It also makes passing the metrics non-optional and tests will fail if the frontend performance changes are greater than 5%.  This could potentially be lowered slightly, but we don't want to throw too many false positives.

<img width="1486" alt="Screenshot 2024-11-07 at 4 29 25 PM" src="https://github.com/user-attachments/assets/d15c43ec-147a-4588-9dc9-700efd96b51c">
<img width="1654" alt="Screenshot 2024-11-07 at 4 29 41 PM" src="https://github.com/user-attachments/assets/3c738fe1-75e9-4d2b-b77a-ee161607aa0c">


**Notes**

* This PR increases the sample size to 11 to lower the variance.  Currently most PRs seem to come in with a +/-2% delta without any changes to performance (due to variables in CPU usage).  
* This PR only checks for frontend server response deltas.  The others have a much larger variance and are not yet reliable.
* This PR removes `cumulativeLayoutShift` from frontend deltas which appeared to be flaky when running a larger sample size

Related to #52343  .



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Navigate to `plugins/woocommerce`
2. Run `pnpm run test:metrics:ci`
3. Check that the deltas are accurately shown for server response time for each test suite (frontend, editor, and product editor)
4. Introduce a slow function like below in `class-woocommerce.php` or checkout this branch `add/52343-pr-alert-test`
5. Run `pnpm run test:metrics:ci`
6. Check that an error is shown if the delta is greater than 5%
7. Check that this PR includes metrics and they are passing
8. Check that the [test PR](https://github.com/woocommerce/woocommerce/pull/52620) includes metrics that are failing

```php
function slow_function() {
	$i = 1;
	while ($i < 5000000) {
		$i++;
	}
}
slow_function();
```



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
